conans/client/source.py

```diff
@@ -87,12 +91,37 @@ def remove_source(raise_error=True):
 
 def config_source_local(dest_dir, conan_file, output):
     output.info('Configuring sources in %s' % dest_dir)
+    conan_file.source_folder = dest_dir
+
     with tools.chdir(dest_dir):
         try:
             with conanfile_exception_formatter(str(conan_file), "source"):
                 with tools.environment_append(conan_file.env):
-                    conan_file.source()
+                    with block_build_and_package_folders():
+                        conan_file.source()
         except ConanExceptionInUserConanfileMethod:
             raise
         except Exception as e:
             raise ConanException(e)
+
+
+@contextmanager
+def block_build_and_package_folders():
+    msg = "Access to '%s' is blocked from source() method: Source method is called once, " \
+          "not one per configuration, so using build_folder or package_folder can produce " \
+          "unexpected results."
+
+    def raise_when_access_build(*args):
+        raise ConanException(msg % "self.build_folder")
+
+    def raise_when_access_package(*args):
+        raise ConanException(msg % "self.package_folder")
+
+    ConanFile.build_folder = property(raise_when_access_build)
```

Comment:
dangearous? Set it at the object level, not the class level. Pass the object as arg.

Reply:
You cannot set a property at object level. But yes, maybe at class level is risky. I'm just assigning it to None  and changing the tests, it was only a little ux improvement in case of bad usage.